capdpa/cxx.py

```diff
@@ -160,60 +289,110 @@ def __convert_namespace(self, cursor):
     def __convert_typedef(self, cursor):
         children = list(cursor.get_children())
         if children:
-            resolved = self.__convert_type(children[0].type)
+            resolved = self.__convert_type([], children[0].type)
         else:
-            resolved = self.__convert_type(cursor.type.get_canonical())
+            resolved = self.__convert_type([], cursor.type.get_canonical())
         return IR.Type_Definition(cursor.type.spelling, resolved)
 
     def __convert_template(self, cursor):
         targs = []
         for c in cursor.get_children():
-            if c.kind == clang.cindex.CursorKind.TEMPLATE_TYPE_PARAMETER:
+            if c.kind in [
+                    clang.cindex.CursorKind.TEMPLATE_TYPE_PARAMETER,
+                    clang.cindex.CursorKind.TEMPLATE_NON_TYPE_PARAMETER]:
                 targs.append(c)
         return IR.Template(
                 entity = IR.Class(
                     name = cursor.spelling,
                     children = self.__convert_children(list(cursor.get_children())[len(targs):])),
-                typenames = [IR.Template_Argument(c.spelling) for c in targs])
+                typenames = [IR.Template_Argument(name = c.spelling,
+                                                  variadic = list(c.get_tokens())[1].spelling == "...") for c in targs])
 
 
     def __convert_children(self, cursors):
         children = []
-        for cursor in cursors:
+        clist = list(cursors)
+        for cursor in clist:
             if cursor.access_specifier in [clang.cindex.AccessSpecifier.PUBLIC, clang.cindex.AccessSpecifier.INVALID]:
-                if cursor.kind == clang.cindex.CursorKind.NAMESPACE:
+                if cursor.semantic_parent and \
+                   cursor.semantic_parent.kind == clang.cindex.CursorKind.CLASS_DECL and \
+                   cursor.lexical_parent and \
+                   cursor.lexical_parent != cursor.semantic_parent:
+                    # Filter out method definitions outside the class declaration
+                    pass
+                elif cursor.kind == clang.cindex.CursorKind.NAMESPACE:
                     children.append(self.__convert_namespace(cursor))
                 elif cursor.kind in [clang.cindex.CursorKind.CLASS_DECL, clang.cindex.CursorKind.STRUCT_DECL]:
-                    children.append(self.__convert_class(cursor))
+                    decl = self.__convert_class(cursor)
```

Comment:
I don't fully understand this code, because you implemented it in 7614f528d4e03f11e742085f65812cce7eca2b16.

Reply:
Oh, I see. Taking a second look I see why this is at this place. This code handles forward declarations. It checks if the declaration already has a definition and if not searches the definition an places it where the forward declaration happened.
This places the definition at the place of its first declaration. This will break if the definition uses something that has been declared later. Yet a forward declaration is usually used to use the defined object before its definition. It's the case I outlined yesterday which causes a circular dependency we cannot resolve anyway. Assuming a forward declaration is only done when already needed there is no way around defining the object at this place.

Reply:
No wonder that code handling forward declarations looks sh*t ;-) Please add a respective comment.